book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+- Range check $r' < 2^t$ (4 rows)
+
+#### Multiplication
+
+In foreign field multiplication, the situation is unfortunately different, and we must check that each of $a, b, q$ and $r$ are less than $f$.  We cannot adopt the strategy from foreign field addition where the operands are allowed to be greater than $f$ because the bit length of $r$ would increases linearly with the number of multiplications.  That is,
+
+$$
+(a_1 + f) \cdot (a_2 + f) = 1 \cdot f + \underbrace{f^2 + (a_1 + a_2 - 1) \cdot f + a_1 \cdot a_2}_{r}
+$$
+
+and after a chain of $k$ multiplication we have
+
+$$
+r = f^k + \ldots + a_1 \cdots a_k
+$$
+
+where $r > f^k$ quickly overflows our CRT modulus $2^t \cdot n$.  For example, assuming our maximum foreign modulus of $f = 2^{259}$ and either of Kimchi's native moduli (i.e. Pallas or Vesta), $f^k > 2^t \cdot n$ for $k > 2$.  That is, an overflow is possible for a chain of greater than 1 foreign field multiplication.  Thus, we must check $a, b, q$ and $r$ are less than $f$ for each multiplication.
+
+Fortunately, in many situations the input operands may already be checked either as inputs or as outputs of previous operations, so they may not be required for each multiplication operation.
+
+Thus, the $q'$ and $r'$ checks (or equivalently $q$ and $r$) are our main focus because they must be done for every multiplication.
+
+- Compute bound $q' = q + f'$ with addition gate (2 rows)
+- Compute bound $r' = r + f'$ with addition gate (2 rows)
+- Range check $q' < 2^t$ (4 rows)
+- Range check $r' < 2^t$ (4 rows)
+
+This costs 12 rows per multiplication.  In a subsequent section, we will reduce it to 8 rows.
+
+### 2-limb decomposition
+
+Due to the limited number of permutable cells per gate, we do not have enough cells for copy constraining $q'$ and $r'$ (or $q$ and $r$) to their respective range check gadgets.  To address this issue, we must decompose $q'$ into 2 limbs instead of 3, like so
+
+$$
+q' = q'_{01} + 2^{2\ell} \cdot q'_2
+$$
+
+and
+
+$$
+q'_{01} = q'_0 + 2^{\ell} \cdot q'_1
+$$
+
+Thus, $q'$ is decomposed into two limbs $q'_{01}$ (at most $2\ell$ bits) and $q'_2$ (at most $\ell$ bits).
+
+Note that $q'$ must be range checked by a `multi-range-check` gadget.  To do this the `multi-range-check` gadget must
+
+- Store a copy of the limbs $q'_0, q'_1$ and $q'_2$ in its witness
+- Range check that they are $\ell$ bit each
+- Constrain that $q'_{01} = q'_0 + 2^{\ell} \cdot q'_1$ (this is done with a special mode of the `multi-range-check` gadget)
+- Have copy constraints for $q'_{01}$ and $q'_2$ as outputs of the `ForeignFieldMul` gate and inputs to the `multi-range-check` gadget
+
+Note that since the foreign field multiplication gate computes $q'$ from $q$ which is already in the witness and $q'_{01}$ and $q'_2$ have copy constraints to a `multi-range-check` gadget that fully constrains their decomposition from $q'$, then the `ForeignFieldMul` gate does not need to store an additional copy of $q'_0$ and $q'_1$.
```

Comment:
Oh so range check does it so ffmul does not have to store more stuff in the layout I see

Reply:
Right!